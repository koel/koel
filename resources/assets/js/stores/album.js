import _ from 'lodash';

import utils from '../services/utils';
import stub from '../stubs/album';
import songStore from './song';

export default {
    stub,
    artists: [],
    
    state: {
        albums: [stub],    
    },

    /**
     * Init the store.
     * 
     * @param  array artists The array of artists to extract album data from.
     */
    init(artists) {
        this.artists = artists;

        // Traverse through the artists array and add their albums into our master album list.
        this.state.albums = _.reduce(artists, (albums, artist) => {
            // While we're doing so, for each album, we get its length
            // and keep a back reference to the artist too.
            _.each(artist.albums, album => {
                album.artist = artist;
                this.getLength(album);
            });

            return albums.concat(artist.albums);
        }, []);

        // Then we init the song store.
        songStore.init(this.state.albums);
    },

    all() {
        return this.state.albums;
    },

    /**
     * Get the total length of an album by summing up its songs' duration.
     * The length will also be converted into a H:i:s format and stored as fmtLength.
     * 
     * @param  object album
     * 
     * @return string The H:i:s format of the album.
     */
    getLength(album) {
        album.length = _.reduce(album.songs, (length, song) => length + song.length, 0);

        return (album.fmtLength = utils.secondsToHis(album.length));
    },
};
