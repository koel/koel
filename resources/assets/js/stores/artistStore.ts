import { reactive, UnwrapNestedRefs } from 'vue'
import { differenceBy, unionBy } from 'lodash'
import { cache, http } from '@/services'
import { arrayify, logger } from '@/utils'

const UNKNOWN_ARTIST_ID = 1
const VARIOUS_ARTISTS_ID = 2

export const artistStore = {
  vault: new Map<number, UnwrapNestedRefs<Artist>>(),

  state: reactive({
    artists: [] as Artist[]
  }),

  byId (id: number) {
    return this.vault.get(id)
  },

  removeByIds (ids: number[]) {
    this.state.artists = differenceBy(this.state.artists, ids.map(id => this.byId(id)), 'id')
    ids.forEach(id => this.vault.delete(id))
  },

  isVarious: (artist: Artist | number) => (typeof artist === 'number')
    ? artist === VARIOUS_ARTISTS_ID
    : artist.id === VARIOUS_ARTISTS_ID,

  isUnknown: (artist: Artist | number) => (typeof artist === 'number')
    ? artist === UNKNOWN_ARTIST_ID
    : artist.id === UNKNOWN_ARTIST_ID,

  isStandard (artist: Artist | number) {
    return !this.isVarious(artist) && !this.isUnknown(artist)
  },

  async uploadImage (artist: Artist, image: string) {
    artist.image = (await http.put<{ imageUrl: string }>(`artist/${artist.id}/image`, { image })).imageUrl

    // sync to vault
    this.byId(artist.id)!.image = artist.image

    return artist.image
  },

  syncWithVault (artists: Artist | Artist[]) {
    return arrayify(artists).map(artist => {
      let local = this.vault.get(artist.id)
      local = local ? Object.assign(local, artist) : reactive(artist)
      this.vault.set(artist.id, local)

      return local
    })
  },

  async resolve (id: number) {
    let artist = this.byId(id)

    if (!artist) {
      try {
        artist = this.syncWithVault(
          await cache.remember<Artist>(['artist', id], async () => await http.get<Artist>(`artists/${id}`))
        )[0]
      } catch (e) {
        logger.error(e)
      }
    }

    return artist
  },

  async paginate (page: number) {
    const resource = await http.get<PaginatorResource>(`artists?page=${page}`)
    this.state.artists = unionBy(this.state.artists, this.syncWithVault(resource.data), 'id')

    return resource.links.next ? ++resource.meta.current_page : null
  }
}
