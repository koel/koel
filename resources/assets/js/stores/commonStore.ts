import isMobile from 'ismobilejs'
import { reactive } from 'vue'
import { http } from '@/services'
import { playlistFolderStore, playlistStore, preferenceStore, queueStore, settingStore, themeStore, userStore } from '.'

interface CommonStoreState {
  allow_download: boolean
  cdn_url: string
  current_user: User
  current_version: string
  latest_version: string
  playlists: Playlist[]
  playlist_folders: PlaylistFolder[]
  settings: Settings
  use_i_tunes: boolean
  use_last_fm: boolean
  use_spotify: boolean
  users: User[]
  use_you_tube: boolean,
  song_count: number,
  song_length: number,
  queue_state: QueueState
}

export const commonStore = {
  state: reactive<CommonStoreState>({
    allow_download: false,
    cdn_url: '',
    current_user: undefined as unknown as User,
    current_version: '',
    latest_version: '',
    playlists: [],
    playlist_folders: [],
    settings: {} as Settings,
    use_i_tunes: false,
    use_last_fm: false,
    use_spotify: false,
    users: [],
    use_you_tube: false,
    song_count: 0,
    song_length: 0,
    queue_state: {
      type: 'queue-states',
      songs: [],
      current_song: null,
      playback_position: 0
    }
  }),

  async init () {
    Object.assign(this.state, await http.get<CommonStoreState>('data'))

    // Always disable YouTube integration on mobile.
    this.state.use_you_tube = this.state.use_you_tube && !isMobile.phone

    // If this is a new user, initialize his preferences to be an empty object.
    this.state.current_user.preferences = this.state.current_user.preferences || {}

    userStore.init(this.state.current_user)
    preferenceStore.init(this.state.current_user)
    playlistStore.init(this.state.playlists)
    playlistFolderStore.init(this.state.playlist_folders)
    settingStore.init(this.state.settings)
    queueStore.init(this.state.queue_state)
    themeStore.init()

    return this.state
  }
}
