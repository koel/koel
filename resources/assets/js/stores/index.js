export * from "./album";
export * from "./artist";
export * from "./equalizer";
export * from "./favorite";
export * from "./playlist";
export * from "./preference";
export * from "./queue";
export * from "./setting";
export * from "./shared";
export * from "./song";
export * from "./user";
