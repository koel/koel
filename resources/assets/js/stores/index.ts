export * from './albumStore'
export * from './artistStore'
export * from './commonStore'
export * from './equalizerStore'
export * from './favoriteStore'
export * from './overviewStore'
export * from './playlistStore'
export * from './playlistFolderStore'
export * from './preferenceStore'
export * from './queueStore'
export * from './recentlyPlayedStore'
export * from './searchStore'
export * from './settingStore'
export * from './songStore'
export * from './themeStore'
export * from './userStore'
export * from './genreStore'
export * from './visualizerStore'
