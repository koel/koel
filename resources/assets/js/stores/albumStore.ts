import { reactive, UnwrapNestedRefs } from 'vue'
import { differenceBy, merge, unionBy } from 'lodash'
import { cache, http } from '@/services'
import { arrayify, logger } from '@/utils'
import { songStore } from '@/stores'

const UNKNOWN_ALBUM_ID = 1

export const albumStore = {
  vault: new Map<number, UnwrapNestedRefs<Album>>(),

  state: reactive({
    albums: [] as Album[]
  }),

  byId (id: number) {
    return this.vault.get(id)
  },

  removeByIds (ids: number[]) {
    this.state.albums = differenceBy(this.state.albums, ids.map(id => this.byId(id)), 'id')
    ids.forEach(id => {
      this.vault.delete(id)
      cache.remove(['album', id])
    })
  },

  isUnknown: (album: Album | number) => {
    if (typeof album === 'number') return album === UNKNOWN_ALBUM_ID
    return album.id === UNKNOWN_ALBUM_ID
  },

  syncWithVault (albums: Album | Album[]) {
    return arrayify(albums).map(album => {
      let local = this.vault.get(album.id)
      local = reactive(local ? merge(local, album) : album)
      this.vault.set(album.id, local)

      return local
    })
  },

  /**
   * Upload a cover for an album.
   *
   * @param {Album} album The album object
   * @param {string} cover The content data string of the cover
   */
  async uploadCover (album: Album, cover: string) {
    album.cover = (await http.put<{ coverUrl: string }>(`album/${album.id}/cover`, { cover })).coverUrl
    songStore.byAlbum(album).forEach(song => song.album_cover = album.cover)

    // sync to vault
    this.byId(album.id)!.cover = album.cover

    return album.cover
  },

  /**
   * Fetch the (blurry) thumbnail-sized version of an album's cover.
   */
  fetchThumbnail: async (id: number) => {
    return (await http.get<{ thumbnailUrl: string }>(`album/${id}/thumbnail`)).thumbnailUrl
  },

  async resolve (id: number) {
    let album = this.byId(id)

    if (!album) {
      try {
        album = this.syncWithVault(
          await cache.remember<Album>(['album', id], async () => await http.get<Album>(`albums/${id}`))
        )[0]
      } catch (e) {
        logger.error(e)
      }
    }

    return album
  },

  async paginate (page: number) {
    const resource = await http.get<PaginatorResource>(`albums?page=${page}`)
    this.state.albums = unionBy(this.state.albums, this.syncWithVault(resource.data), 'id')

    return resource.links.next ? ++resource.meta.current_page : null
  },

  async fetchForArtist (artist: Artist | number) {
    const id = typeof artist === 'number' ? artist : artist.id

    return this.syncWithVault(
      await cache.remember<Album[]>(['artist-albums', id], async () => await http.get<Album[]>(`artists/${id}/albums`))
    )
  }
}
