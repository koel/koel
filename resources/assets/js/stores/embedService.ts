import { http } from '@/services/http'
import { themeStore } from '@/stores/themeStore'

export const embedService = {
  async resolveForEmbeddable (embeddable: Embeddable) {
    let embeddableType = embeddable.type.slice(0, -1) // remove the last 's' to get the singular form

    if (embeddableType === 'song' || embeddableType === 'episode') {
      embeddableType = 'playable'
    }

    return await http.post<Embed>('embeds/resolve', {
      embeddable_id: embeddable.id,
      embeddable_type: embeddableType,
    })
  },

  async encryptOptions (options: EmbedOptions) {
    const { encrypted } = await http.post<{ encrypted: string }>('embed-options', {
      ...options,
    })

    return encrypted
  },

  async getWidgetPayload (id: string, encryptedOptions: string) {
    const payload = await http.get<{ embed: WidgetReadyEmbed, options: EmbedOptions, theme: Theme | null }>(
      `embeds/${id}/${encryptedOptions}`,
    )

    // Since the theme value can be any literal string, do a sanity check.
    payload.options.theme = themeStore.isValidTheme(payload.options.theme)
      ? payload.options.theme
      : themeStore.getDefaultTheme().id

    return payload
  },
}
