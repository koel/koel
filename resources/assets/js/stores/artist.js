import _ from 'lodash';

import config from '../config';
import albumStore from './album';
import sharedStore from './shared';

export default {
    state: {
        artists: [],
    },

    /**
     * Init the store.
     * 
     * @param  array artists The array of artists we got from the server.
     */
    init(artists = null) {
        this.state.artists = artists ? artists: sharedStore.state.artists;

        // Init the album store. This must be called prior to the next logic,
        // because we're using some data from the album store later.
        albumStore.init(this.state.artists);

        // Traverse through artists array to get the cover and number of songs for each.
        _.each(this.state.artists, artist => {
            this.getCover(artist);
            
            artist.songCount = _.reduce(artist.albums, (count, album)  => count + album.songs.length, 0);
        });
    },

    all() {
        return this.state.artists;
    },

    /**
     * Get all songs performed by an artist.
     *
     * @param object artist
     *
     * @return array
     */
    getSongsByArtist(artist) {
        if (!artist.songs) {
            artist.songs = _.reduce(artist.albums, (songs, album) => songs.concat(album.songs), []);
        }

        return artist.songs;
    },

    /**
     * Get the artist's cover
     *
     * @param object artist
     *
     * @return string
     */
    getCover(artist) {
        artist.cover = config.unknownCover;

        artist.albums.every(album => {
            // If there's a "real" cover, use it.
            if (album.cover != config.unknownCover) {
                artist.cover = album.cover;
                
                // I want to break free.
                return false;
            }
        });

        return artist.cover;
    },
};
