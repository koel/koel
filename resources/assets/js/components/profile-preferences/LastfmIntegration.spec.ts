import { describe, expect, it } from 'vitest'
import { createHarness } from '@/__tests__/TestHarness'
import { commonStore } from '@/stores/commonStore'
import Component from './LastfmIntegration.vue'

describe('lastfmIntegration.vue', () => {
  const h = createHarness()

  it.each<[boolean, boolean]>([[false, false], [false, true], [true, false], [true, true]])(
    'renders proper content with Last.fm integration status %s, current user admin status %s',
    (useLastfm, isAdmin) => {
      commonStore.state.uses_last_fm = useLastfm

      if (isAdmin) {
        h.beAdmin()
      } else {
        h.be()
      }

      expect(h.render(Component).html()).toMatchSnapshot()
    },
  )
})
