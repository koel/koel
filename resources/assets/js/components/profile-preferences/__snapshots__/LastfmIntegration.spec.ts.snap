// Vitest Snapshot v1

exports[`lastfmIntegration.vue > renders proper content with Last.fm integration status false, current user admin status false 1`] = `
<section data-v-6bdc1e9f="" class="text-k-text-secondary">
  <h3 data-v-6bdc1e9f="" class="text-2xl mb-2"><span data-v-6bdc1e9f="" class="mr-2 text-[var(--lastfm-color)]"><br data-v-6bdc1e9f="" data-testid="Icon" icon="[object Object]"></span> Last.fm Integration </h3>
  <div data-v-6bdc1e9f="" data-testid="lastfm-not-integrated">
    <p data-v-6bdc1e9f=""> Last.fm integration is not enabled. <span data-v-6bdc1e9f="" data-testid="lastfm-user-instruction"> Try politely asking an administrator to enable it. </span></p>
  </div>
</section>
`;

exports[`lastfmIntegration.vue > renders proper content with Last.fm integration status false, current user admin status true 1`] = `
<section data-v-6bdc1e9f="" class="text-k-text-secondary">
  <h3 data-v-6bdc1e9f="" class="text-2xl mb-2"><span data-v-6bdc1e9f="" class="mr-2 text-[var(--lastfm-color)]"><br data-v-6bdc1e9f="" data-testid="Icon" icon="[object Object]"></span> Last.fm Integration </h3>
  <div data-v-6bdc1e9f="" data-testid="lastfm-not-integrated">
    <p data-v-6bdc1e9f=""> Last.fm integration is not enabled. <span data-v-6bdc1e9f="" data-testid="lastfm-admin-instruction"> Check <a data-v-6bdc1e9f="" href="https://docs.koel.dev/service-integrations#last-fm" target="_blank">Documentation</a> for integration instructions. </span></p>
  </div>
</section>
`;

exports[`lastfmIntegration.vue > renders proper content with Last.fm integration status true, current user admin status false 1`] = `
<section data-v-6bdc1e9f="" class="text-k-text-secondary">
  <h3 data-v-6bdc1e9f="" class="text-2xl mb-2"><span data-v-6bdc1e9f="" class="mr-2 text-[var(--lastfm-color)]"><br data-v-6bdc1e9f="" data-testid="Icon" icon="[object Object]"></span> Last.fm Integration </h3>
  <div data-v-6bdc1e9f="" data-testid="lastfm-integrated">
    <p data-v-6bdc1e9f="">Last.fm integration is enabled. Koel will attempt to retrieve album and artist information from Last.fm.</p>
    <p data-v-6bdc1e9f=""> It appears that you have connected your Last.fm account as well – Perfect! </p>
    <p data-v-6bdc1e9f=""> Connecting Koel and your Last.fm account enables such exciting features as <a data-v-6bdc1e9f="" href="https://www.last.fm/about/trackmymusic" rel="noopener" target="_blank">scrobbling</a>. </p>
    <div data-v-6bdc1e9f="" class="buttons mt-4 space-x-2">
      <!---->
      <!---->
    </div>
  </div>
</section>
`;

exports[`lastfmIntegration.vue > renders proper content with Last.fm integration status true, current user admin status true 1`] = `
<section data-v-6bdc1e9f="" class="text-k-text-secondary">
  <h3 data-v-6bdc1e9f="" class="text-2xl mb-2"><span data-v-6bdc1e9f="" class="mr-2 text-[var(--lastfm-color)]"><br data-v-6bdc1e9f="" data-testid="Icon" icon="[object Object]"></span> Last.fm Integration </h3>
  <div data-v-6bdc1e9f="" data-testid="lastfm-integrated">
    <p data-v-6bdc1e9f="">Last.fm integration is enabled. Koel will attempt to retrieve album and artist information from Last.fm.</p>
    <p data-v-6bdc1e9f=""> It appears that you have connected your Last.fm account as well – Perfect! </p>
    <p data-v-6bdc1e9f=""> Connecting Koel and your Last.fm account enables such exciting features as <a data-v-6bdc1e9f="" href="https://www.last.fm/about/trackmymusic" rel="noopener" target="_blank">scrobbling</a>. </p>
    <div data-v-6bdc1e9f="" class="buttons mt-4 space-x-2">
      <!---->
      <!---->
    </div>
  </div>
</section>
`;
